File: react-api/db/crud.ts
import { TodoItem } from "../models/TodoItem";
import { connect } from "./db";

export async function create(todoItem: TodoItem): Promise<number> {
  const db = await connect();

  const stmt = await db.prepare(
    "INSERT INTO TODOs (title, completed) VALUES (?, ?)"
  );

  const result = await stmt.run(todoItem.title, todoItem.completed ? 1 : 0);

  const { lastID } = result;

  if (!lastID) throw new Error("Could not create record");

  await stmt.finalize();

  return lastID;
}

export async function readAll(): Promise<TodoItem[]> {
  const db = await connect();

  const results = await db.all("SELECT * FROM TODOs");

  return results.map(
    (result): TodoItem => ({
      id: result.id,
      title: result.title,
      completed: result.completed === 1,
    })
  );
}

export async function update(id: number, todo: TodoItem): Promise<number> {
  const db = await connect();

  const stmt = await db.prepare(
    "UPDATE TODOs SET title = ?, completed = ? WHERE id = ?"
  );

  const result = await stmt.run(todo.title, todo.completed ? 1 : 0, id);

  const { lastID } = result;

  if (!lastID) throw new Error("Could not create record");

  await stmt.finalize();

  return lastID;
}

File: htmx-python-todo/htmx_python_todo/crud.py
import os
from .db.db import get_connection
from .models import TodoItem


def create(item: TodoItem):
    with get_connection() as cur:
        sql = "INSERT INTO TODOs (title, completed) VALUES (?, ?)"
        cur.execute(sql, (item.title, item.completed))


def readAll():
    with get_connection() as db:
        rows = db.execute("SELECT * FROM TODOs").fetchall()
    return [TodoItem(itemId=row[0], title=row[1], completed=row[2]) for row in rows]


def update(rowId: int, item: TodoItem):
    with get_connection() as cur:
        sql = "UPDATE TODOs SET title = ?, completed = ? WHERE id = ?"
        cur.execute(sql, (item.title, item.completed, rowId))

def delRecord(rowId: int):
    with get_connection() as cur:
        sql = "DELETE FROM TODOs WHERE id = ?"
        cur.execute(sql, (rowId,))
